File: day6/day6.exs
{:ok, file} = File.read("6")
inputs = String.split(file, "\n")
nums = Enum.map(Enum.slice(inputs, 0..3), fn x -> String.graphemes(x) end)
zipped = Enum.zip_with(nums, fn [w, x, y, z] -> [w, x, y, z] end)
ops = String.split(Enum.at(inputs, 4))

defmodule Part2 do
  def solve([], [], acc) do
    acc
  end

  def solve([nums | rest_nums], [op | rest_ops], acc) do
    if op == "*" do
      solve(rest_nums, rest_ops, acc + Enum.reduce(nums, fn x, y -> x * y end))
    else
      solve(rest_nums, rest_ops, acc + Enum.sum(nums))
    end
  end

  def parse([], acc) do
    Enum.reverse(acc)
  end

  def parse([head | tail], [acc_head | acc_tail]) do
    if head == [" ", " ", " ", " "] do
      parse(tail, [[] | [acc_head | acc_tail]])
    else
      parsed = String.to_integer(String.trim(List.to_string(head)))
      parse(tail, [[parsed | acc_head] | acc_tail])
    end
  end

  def parse(nums) do
    parse(nums, [[]])
  end
end

IO.puts(Part2.solve(Part2.parse(zipped), ops, 0))

File: day6/day6.py
from functools import reduce

inp = open("6").read().splitlines()
nums = [list(map(int, i.split())) for i in inp[:-1]]
ops = inp[-1].split()

mult = lambda x, y: x * y
add = lambda x, y: x + y

part1 = 0

for i in range(len(ops)):
    part1 += reduce(
        mult if ops[i] == "*" else add,
        [nums[j][i] for j in range(len(nums))],
        1 if ops[i] == "*" else 0,
    )

for i in range(len(inp)):
    inp[i] += " "


col_index = 0
col_nums = []
part2 = 0

for i in range(len(inp[0])):
    col = "".join([inp[j][i] for j in range(len(nums))])
    if col.strip() != "":
        num = int("".join([inp[j][i] for j in range(len(nums))]))
        col_nums.append(num)
    else:
        part2 += reduce(
            mult if ops[col_index] == "*" else add,
            col_nums,
            1 if ops[col_index] == "*" else 0,
        )
        col_index += 1
        col_nums = []

print(part1)
print(part2)
